### PR TITLE
[Clang] Improve type traits recognition in `__has_builtin`

### LINKED ISSUE #111477: __has_builtin(__reference_binds_to_temporary) became false in Clang 19
Likewise for `__has_builtin(__reference_constructs_from_temporary)`. 

Is that intentional? What's the replacement if I want to detect whether the compiler supports them?

### PR DESCRIPTION
**Pull request by cor3ntin**
`__has_builtin` was relying on reversible identifiers and string matching to recognize builtin-type traits, leading to some newer type traits not being recognized.

Fixes #111477

### CODE REVIEW

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -635,8 +637,8 @@ New features
   if class of allocation and deallocation function mismatches.
   `Documentation <https://clang.llvm.org/docs/analyzer/checkers.html#unix-mismatcheddeallocator-c-c>`__.
 
-- Function effects, e.g. the ``nonblocking`` and ``nonallocating`` "performance constraint" 
-  attributes, are now verified. For example, for functions declared with the ``nonblocking`` 
+- Function effects, e.g. the ``nonblocking`` and ``nonallocating`` "performance constraint"
```
Comment (erichkeane) [inline clang/docs/ReleaseNotes.rst:640]: unrelated :(

**Review by erichkeane — COMMENTED** [review]
This seems reasonable to me.  Its a shame that TRANSFORM_TYPE_TRAIT_DEF can't be done in terms of TYPE_TRAIT, but otherwiseI think this is a good direction.

**Review by erichkeane — APPROVED** [review]

File: clang/test/Preprocessor/feature_tests.cpp [hunk 2]
```diff
@@ -31,7 +31,10 @@
     !__has_builtin(__underlying_type) || \
     !__has_builtin(__is_trivial) || \
     !__has_builtin(__is_same_as) || \
-    !__has_builtin(__has_unique_object_representations)
+    !__has_builtin(__has_unique_object_representations) || \
+    !__has_builtin(__is_trivially_equality_comparable) || \
+    !__has_builtin(__reference_constructs_from_temporary) || \
```
Comment (timsong-cpp) [inline clang/test/Preprocessor/feature_tests.cpp:36]: `__reference_converts_from_temporary` too?

**Review by timsong-cpp — COMMENTED** [review]
Can a fix be backported to 19?
